v2/spiffetls/spiffetls_test.go

```diff
@@ -0,0 +1,470 @@
+			serverConnPeerIDErr: "spiffetls: no peer certificates",
+		},
+		{
+			name:       "MTLSClientWithRawConfig / MTLSServerWithRawConfig succeeds",
+			dialMode:   spiffetls.MTLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), svidSource, bundleSource),
+			listenMode: spiffetls.MTLSServerWithRawConfig(tlsconfig.AuthorizeID(clientID), wlAPISourceA, bundleSource),
+		},
+		{
+			name:                "MTLSWebClientWithRawConfig / MTLSWebServerWithRawConfig succeeds",
+			dialMode:            spiffetls.MTLSWebClientWithRawConfig(webCertPool, svidSource),
+			listenMode:          spiffetls.MTLSWebServerWithRawConfig(tlsconfig.AuthorizeID(clientID), webCert, bundleSource),
+			clientConnPeerIDErr: "spiffetls: no URI SANs",
+		},
+	}
+
+	testClose(t, spiffetls.TLSServerWithSource(wlAPISourceA), spiffetls.TLSClientWithSource(tlsconfig.AuthorizeID(serverID), wlAPISourceB))
+
+	for _, test := range tests {
+		test := test
+
+		if test.defaultWlAPIAddr != "" {
+			require.NoError(t, os.Setenv("SPIFFE_ENDPOINT_SOCKET", test.defaultWlAPIAddr))
+		} else {
+			require.NoError(t, os.Unsetenv("SPIFFE_ENDPOINT_SOCKET"))
+		}
+
+		t.Run(test.name, func(t *testing.T) {
+			// Start listening
+			listenCtx, cancelListenCtx := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancelListenCtx()
+
+			var wg sync.WaitGroup
+			var listener net.Listener
+			var listenAddr string
+			listenDataCh := make(chan string, 1)
+			listenErrCh := make(chan error, 1)
+			if test.listenLAddr == "" {
+				test.listenLAddr = defaultListenLAddr
+			}
+			if test.listenProtocol == "" {
+				test.listenProtocol = defaultListenProtocol
+			}
+			if test.listenMode != nil {
+				listener, err = spiffetls.ListenWithMode(listenCtx, test.listenProtocol, test.listenLAddr, test.listenMode, test.listenOption...)
+				if test.listenErr != "" {
+					require.EqualError(t, err, test.listenErr)
+					return
+				}
+				require.NoError(t, err)
+				require.NotNil(t, listener)
+				defer listener.Close()
+
+				listenAddr = listener.Addr().String()
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					conn, err := listener.Accept()
+					if err != nil {
+						listenErrCh <- err
+						return
+					}
+					defer conn.Close()
+
+					data, err := bufio.NewReader(conn).ReadString('\n')
+					if err != nil {
+						listenErrCh <- err
+						return
+					}
+					listenDataCh <- data
+
+					// Test serverConn.PeerID()
+					spiffeID, err := spiffetls.PeerIDFromConn(conn)
+					if test.serverConnPeerIDErr == "" {
+						require.NoError(t, err)
+						require.Equal(t, clientID, spiffeID)
+					} else {
+						require.EqualError(t, err, test.serverConnPeerIDErr)
+					}
+				}()
+			} else {
+				// Test Listen function
+				listener, err = spiffetls.Listen(listenCtx, test.listenProtocol, test.listenLAddr, tlsconfig.AuthorizeID(clientID))
+				if test.listenErr != "" {
+					require.EqualError(t, err, test.listenErr)
```

Comment:
We shouldn't call `require.*` or other testify packages that ends up calling to `t.FailNow()` inside go-routines.
```
FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test.
```
https://golang.org/pkg/testing/#T.FailNow

As an alternative we could exfiltrate the error through a channel and make the assertion in the `select` below.

Reply:
Yes, good catch.

Reply:
What if those `require.*` inside go routines are replaced with `assert.` calls? I went through them and continue the execution after the assertion failures shouldn't cause a panic. I'm just trying to think an alternative to not add extra complexity to the code.

Reply:
Asserts from other goroutines are ok as long as they don't happen after the test method has finished (the testing package complains). So as long as the test waits for those goroutines to finish before returning, it should be alright.